File: taurus-protect-sdk-go/pkg/protect/mapper/price.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// PriceFromDTO converts an OpenAPI CurrencyPrice to a domain Price.
func PriceFromDTO(dto *openapi.TgvalidatordCurrencyPrice) *model.Price {
	if dto == nil {
		return nil
	}

	price := &model.Price{
		Blockchain:          safeString(dto.Blockchain),
		CurrencyFrom:        safeString(dto.CurrencyFrom),
		CurrencyTo:          safeString(dto.CurrencyTo),
		Decimals:            safeString(dto.Decimals),
		Rate:                safeString(dto.Rate),
		ChangePercent24Hour: safeString(dto.ChangePercent24Hour),
		Source:              safeString(dto.Source),
		CurrencyFromInfo:    CurrencyInfoFromDTO(dto.CurrencyFromInfo),
		CurrencyToInfo:      CurrencyInfoFromDTO(dto.CurrencyToInfo),
	}

	// Convert timestamps
	if dto.CreationDate != nil {
		price.CreationDate = *dto.CreationDate
	}
	if dto.UpdateDate != nil {
		price.UpdateDate = *dto.UpdateDate
	}

	// Convert signatures
	if dto.Signatures != nil {
		price.Signatures = PriceSignaturesFromDTO(dto.Signatures)
	}

	return price
}

// PricesFromDTO converts a slice of OpenAPI CurrencyPrice to domain Prices.
func PricesFromDTO(dtos []openapi.TgvalidatordCurrencyPrice) []*model.Price {
	if dtos == nil {
		return nil
	}
	prices := make([]*model.Price, len(dtos))
	for i := range dtos {
		prices[i] = PriceFromDTO(&dtos[i])
	}
	return prices
}

// PriceSignatureFromDTO converts an OpenAPI CurrencyPriceSignature to a domain PriceSignature.
func PriceSignatureFromDTO(dto *openapi.TgvalidatordCurrencyPriceSignature) model.PriceSignature {
	if dto == nil {
		return model.PriceSignature{}
	}
	return model.PriceSignature{
		UserID:    dto.UserId,
		Signature: dto.Signature,
	}
}

// PriceSignaturesFromDTO converts a slice of OpenAPI CurrencyPriceSignature to domain PriceSignatures.
func PriceSignaturesFromDTO(dtos []openapi.TgvalidatordCurrencyPriceSignature) []model.PriceSignature {
	if dtos == nil {
		return nil
	}
	signatures := make([]model.PriceSignature, len(dtos))
	for i := range dtos {
		signatures[i] = PriceSignatureFromDTO(&dtos[i])
	}
	return signatures
}

// PriceHistoryPointFromDTO converts an OpenAPI PricesHistoryPoint to a domain PriceHistoryPoint.
func PriceHistoryPointFromDTO(dto *openapi.TgvalidatordPricesHistoryPoint) *model.PriceHistoryPoint {
	if dto == nil {
		return nil
	}

	point := &model.PriceHistoryPoint{
		Blockchain:       safeString(dto.Blockchain),
		CurrencyFrom:     safeString(dto.CurrencyFrom),
		CurrencyTo:       safeString(dto.CurrencyTo),
		High:             safeString(dto.High),
		Low:              safeString(dto.Low),
		Open:             safeString(dto.Open),
		Close:            safeString(dto.Close),
		VolumeFrom:       safeString(dto.VolumeFrom),
		VolumeTo:         safeString(dto.VolumeTo),
		ChangePercent:    safeString(dto.ChangePercent),
		CurrencyFromInfo: CurrencyInfoFromDTO(dto.CurrencyFromInfo),
		CurrencyToInfo:   CurrencyInfoFromDTO(dto.CurrencyToInfo),
	}

	// Convert timestamp
	if dto.PeriodStartDate != nil {
		point.PeriodStartDate = *dto.PeriodStartDate
	}

	return point
}

// PriceHistoryPointsFromDTO converts a slice of OpenAPI PricesHistoryPoint to domain PriceHistoryPoints.
func PriceHistoryPointsFromDTO(dtos []openapi.TgvalidatordPricesHistoryPoint) []*model.PriceHistoryPoint {
	if dtos == nil {
		return nil
	}
	points := make([]*model.PriceHistoryPoint, len(dtos))
	for i := range dtos {
		points[i] = PriceHistoryPointFromDTO(&dtos[i])
	}
	return points
}

// ConversionValueFromDTO converts an OpenAPI ConversionValue to a domain ConversionValue.
func ConversionValueFromDTO(dto *openapi.TgvalidatordConversionValue) *model.ConversionValue {
	if dto == nil {
		return nil
	}
	return &model.ConversionValue{
		Symbol:        safeString(dto.Symbol),
		Value:         safeString(dto.Value),
		MainUnitValue: safeString(dto.MainUnitValue),
		CurrencyInfo:  CurrencyInfoFromDTO(dto.CurrencyInfo),
	}
}

// ConversionValuesFromDTO converts a slice of OpenAPI ConversionValue to domain ConversionValues.
func ConversionValuesFromDTO(dtos []openapi.TgvalidatordConversionValue) []*model.ConversionValue {
	if dtos == nil {
		return nil
	}
	values := make([]*model.ConversionValue, len(dtos))
	for i := range dtos {
		values[i] = ConversionValueFromDTO(&dtos[i])
	}
	return values
}

// ConversionResultFromDTO converts an OpenAPI ConversionReply to a domain ConversionResult.
func ConversionResultFromDTO(dto *openapi.TgvalidatordConversionReply) *model.ConversionResult {
	if dto == nil {
		return nil
	}
	return &model.ConversionResult{
		CurrencyFrom:     safeString(dto.CurrencyFrom),
		BaseCurrency:     safeString(dto.BaseCurrency),
		Values:           ConversionValuesFromDTO(dto.Result),
		FullCurrencyFrom: CurrencyInfoFromDTO(dto.FullCurrencyFrom),
		FullBaseCurrency: CurrencyInfoFromDTO(dto.FullBaseCurrency),
	}
}

File: taurus-protect-sdk-typescript/src/mappers/price.ts
/**
 * Price mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { ConversionResult, Price, PriceHistoryPoint } from '../models/price';
import { currencyFromDto } from './currency';
import { safeDate, safeMap, safeString } from './base';

/**
 * Maps a price DTO to a Price domain model.
 */
export function priceFromDto(dto: unknown): Price | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    blockchain: safeString(d.blockchain),
    currencyFrom: safeString(d.currencyFrom ?? d.currency_from),
    currencyTo: safeString(d.currencyTo ?? d.currency_to),
    decimals: safeString(d.decimals),
    rate: safeString(d.rate),
    changePercent24Hour: safeString(d.changePercent24Hour ?? d.change_percent_24_hour),
    source: safeString(d.source),
    // Map creationDate -> createdAt, updateDate -> updatedAt (as in Java mapper)
    createdAt: safeDate(d.creationDate ?? d.creation_date ?? d.createdAt ?? d.created_at),
    updatedAt: safeDate(d.updateDate ?? d.update_date ?? d.updatedAt ?? d.updated_at),
    currencyFromInfo: currencyFromDto(d.currencyFromInfo ?? d.currency_from_info),
    currencyToInfo: currencyFromDto(d.currencyToInfo ?? d.currency_to_info),
  };
}

/**
 * Maps an array of price DTOs to Price domain models.
 */
export function pricesFromDto(dtos: unknown[] | null | undefined): Price[] {
  return safeMap(dtos, priceFromDto);
}

/**
 * Maps a price history point DTO to a PriceHistoryPoint domain model.
 */
export function priceHistoryPointFromDto(dto: unknown): PriceHistoryPoint | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    periodStartDate: safeDate(d.periodStartDate ?? d.period_start_date),
    blockchain: safeString(d.blockchain),
    currencyFrom: safeString(d.currencyFrom ?? d.currency_from),
    currencyTo: safeString(d.currencyTo ?? d.currency_to),
    high: safeString(d.high),
    low: safeString(d.low),
    open: safeString(d.open),
    close: safeString(d.close),
    volumeFrom: safeString(d.volumeFrom ?? d.volume_from),
    volumeTo: safeString(d.volumeTo ?? d.volume_to),
    changePercent: safeString(d.changePercent ?? d.change_percent),
    currencyFromInfo: currencyFromDto(d.currencyFromInfo ?? d.currency_from_info),
    currencyToInfo: currencyFromDto(d.currencyToInfo ?? d.currency_to_info),
  };
}

/**
 * Maps an array of price history point DTOs to PriceHistoryPoint domain models.
 */
export function priceHistoryPointsFromDto(dtos: unknown[] | null | undefined): PriceHistoryPoint[] {
  return safeMap(dtos, priceHistoryPointFromDto);
}

/**
 * Maps a conversion value DTO to a ConversionResult domain model.
 */
export function conversionResultFromDto(dto: unknown): ConversionResult | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    symbol: safeString(d.symbol),
    value: safeString(d.value),
    mainUnitValue: safeString(d.mainUnitValue ?? d.main_unit_value),
    currencyInfo: currencyFromDto(d.currencyInfo ?? d.currency_info),
  };
}

/**
 * Maps an array of conversion value DTOs to ConversionResult domain models.
 */
export function conversionResultsFromDto(dtos: unknown[] | null | undefined): ConversionResult[] {
  return safeMap(dtos, conversionResultFromDto);
}
